packages/lib-core/src/runtime/PluginLoader.ts

```diff
@@ -0,0 +1,553 @@
+import type { AnyObject } from '@monorepo/common';
```

Comment:
@vojtechszocs Github couldn't diff these files for some reason (this moved from `store/PluginLoader.ts` to here). I diffed the two locally:

```diff
2c2
< import { consoleLogger } from '@monorepo/common';
---
> import { consoleLogger, ErrorWithCause } from '@monorepo/common';
5c5
< import { PLUGIN_MANIFEST, REMOTE_ENTRY_SCRIPT, REMOTE_ENTRY_CALLBACK } from '../constants';
---
> import { DEFAULT_PLUGIN_MANIFEST, DEFAULT_REMOTE_ENTRY_CALLBACK } from '../constants';
9a10,12
> import { settleAllPromises } from '../utils/promise';
> import { getRandomString } from '../utils/random';
> import { injectScriptElement, getScriptElement } from '../utils/scripts';
12a16,21
> declare global {
>   interface Window {
>     [name: string]: PluginEntryCallback;
>   }
> }
> 
14d22
<   entryCallbackFired: boolean;
16a25,26
>   entryCallbackFired?: boolean;
>   entryCallbackModule?: PluginEntryModule;
45,47c55,79
<   /** Control which plugins can be loaded. */
<   canLoadPlugin: (pluginName: string) => boolean;
<   /** Name of the global callback function used by plugin entry scripts. */
---
>   /**
>    * Control which plugins can be loaded.
>    *
>    * The `reload` argument indicates whether an already loaded plugin is
>    * to be reloaded, assuming there was a change in the plugin manifest.
>    *
>    * By default, all plugins are allowed to be loaded.
>    */
>   canLoadPlugin: (manifest: PluginManifest, reload: boolean) => boolean;
> 
>   /**
>    * Control whether the given plugin script can be reloaded when attempting
>    * to reload the associated plugin.
>    *
>    * By default, all plugin scripts are allowed to be reloaded.
>    */
>   canReloadScript: (manifest: PluginManifest, scriptName: string) => boolean;
> 
>   /**
>    * Name of the global callback function used by plugin entry scripts.
>    *
>    * Applies only to plugins using the `callback` registration method.
>    *
>    * Default value: `__load_plugin_entry__`.
>    */
49c81,90
<   /** Custom resource fetch implementation. */
---
> 
>   /**
>    * Custom resource fetch implementation.
>    *
>    * The custom implementation may specify any host application or environment
>    * specific request headers that are necessary to fetch plugin resources over
>    * the network.
>    *
>    * By default, a basic {@link fetch} API based implementation is used.
>    */
51c92,106
<   /** Fixed resolutions used when processing plugin dependencies. */
---
> 
>   /**
>    * Fixed resolutions used when processing plugin dependencies.
>    *
>    * There are two kinds of dependencies:
>    * - Regular dependencies, which refer to actual plugins that must be loaded with
>    *   the right version prior to loading the plugin that depends on them.
>    * - Environment dependencies, which refer to any environment specific constants
>    *   such as the host application's version, runtime platform version, etc.
>    *
>    * This option should be used to satisfy any environment specific dependencies
>    * supported by the host application.
>    *
>    * Default value: empty object.
>    */
53c108,118
<   /** Shared scope object for initializing `PluginEntryModule` containers. */
---
> 
>   /**
>    * Shared scope object for initializing `PluginEntryModule` containers.
>    *
>    * Host applications built with webpack should use dedicated webpack specific APIs
>    * such as `__webpack_init_sharing__` and `__webpack_share_scopes__`.
>    *
>    * Default value: empty object.
>    *
>    * @see https://webpack.js.org/concepts/module-federation/#dynamic-remote-containers
>    */
55,56c120,138
<   /** Post-process the plugin manifest. Can be used as a custom validation hook. */
<   postProcessManifest: (manifest: PluginManifest) => Promise<PluginManifest>;
---
> 
>   /**
>    * Post-process the plugin manifest when fetched over the network.
>    *
>    * By default, no post-processing is performed on the manifest.
>    */
>   postProcessManifest: (manifest: PluginManifest) => PluginManifest;
> 
>   /**
>    * Provide access to the plugin's entry module.
>    *
>    * Applies only to plugins using the `custom` registration method.
>    *
>    * For example, if a plugin was built with `var` library type (i.e. its entry module
>    * is assigned to a variable), you can access the entry module via `window[pluginName]`.
>    *
>    * By default, this function does nothing.
>    */
>   getPluginEntryModule: (manifest: PluginManifest) => PluginEntryModule | void;
74c156
<       entryCallbackName: options.entryCallbackName ?? REMOTE_ENTRY_CALLBACK,
---
>       canReloadScript: options.canReloadScript ?? (() => true),
78c160,162
<       postProcessManifest: options.postProcessManifest ?? (async (manifest) => manifest),
---
>       postProcessManifest: options.postProcessManifest ?? _.identity,
>       entryCallbackName: options.entryCallbackName ?? DEFAULT_REMOTE_ENTRY_CALLBACK,
>       getPluginEntryModule: options.getPluginEntryModule ?? _.noop,
108c192
<    * Start loading a plugin from `baseURL`.
---
>    * Load a plugin from the given URL.
110,111c194,195
<    * This involves the following steps:
<    * - get plugin manifest
---
>    * This involves the following asynchronous operations:
>    * - load plugin manifest (unless provided as an object)
113c197
<    * - load plugin entry script
---
>    * - load plugin scripts
115,116c199,201
<    * The last step delegates control to the plugin entry script, which is supposed to call
<    * the global callback function created via {@link registerPluginEntryCallback} method.
---
>    * Plugins using the `callback` registration method are expected to call the global entry
>    * callback function created via {@link registerPluginEntryCallback} method, passing two
>    * arguments: plugin name and the entry module.
117a203,206
>    * For plugins using the `custom` registration method, the `getPluginEntryModule` function
>    * is expected to return the entry module of the given plugin. If not implemented properly,
>    * plugins using the `custom` registration method will fail to load.
>    *
120c209,212
<   async loadPlugin(baseURL: string) {
---
>   async loadPlugin(
>     baseURL: string,
>     manifestNameOrObject: string | PluginManifest = DEFAULT_PLUGIN_MANIFEST,
>   ) {
124c216,219
<       manifest = await this.getPluginManifest(baseURL);
---
>       manifest =
>         typeof manifestNameOrObject === 'string'
>           ? await this.loadPluginManifest(resolveURL(baseURL, manifestNameOrObject))
>           : manifestNameOrObject;
128c223
<         errorMessage: `Failed to get a valid plugin manifest from ${baseURL}`,
---
>         errorMessage: 'Failed to load plugin manifest',
133a229,240
>     try {
>       manifest = pluginManifestSchema.strict(true).validateSync(manifest, { abortEarly: false });
>     } catch (e) {
>       this.invokeListeners({
>         success: false,
>         pluginName: typeof manifest?.name === 'string' ? manifest.name : undefined,
>         errorMessage: 'Failed to validate plugin manifest',
>         errorCause: e,
>       });
>       return;
>     }
> 
134a242
>     const reload = this.plugins.has(pluginName);
137c245
<       consoleLogger.warn(`Attempt to reload plugin ${pluginName} while being loaded`);
---
>       consoleLogger.warn(`Attempt to reload plugin ${pluginName} which has not finished loading`);
141,142c249,254
<     if (this.plugins.get(pluginName)?.status === 'loaded') {
<       consoleLogger.warn(`Attempt to reload plugin ${pluginName} after being loaded`);
---
>     if (
>       reload &&
>       this.plugins.get(pluginName)?.status === 'loaded' &&
>       _.isEqual(manifest.compilationHash, this.plugins.get(pluginName)?.manifest.compilationHash)
>     ) {
>       consoleLogger.warn(`Attempt to reload plugin ${pluginName} with same compilation hash`);
146,150c258
<     const data: PluginLoadData = {
<       entryCallbackFired: false,
<       status: 'pending',
<       manifest,
<     };
---
>     const data: PluginLoadData = { status: 'pending', manifest };
151a260,263
>     if (manifest.registrationMethod === 'callback') {
>       data.entryCallbackFired = false;
>     }
> 
154c266
<     if (!this.options.canLoadPlugin(pluginName)) {
---
>     if (!this.options.canLoadPlugin(manifest, reload)) {
159c271
<         errorMessage: `Plugin ${pluginName} is not allowed to be loaded`,
---
>         errorMessage: `Plugin ${pluginName} is not allowed to be ${reload ? 'reloaded' : 'loaded'}`,
178c290
<       this.loadPluginEntryScript(baseURL, manifest);
---
>       await this.loadPluginScripts(baseURL, manifest, data);
184c296
<         errorMessage: `Failed to start loading entry script of plugin ${pluginName}`,
---
>         errorMessage: `Failed to load scripts of plugin ${pluginName}`,
191c303
<    * Fetch the plugin manifest from `baseURL` and validate it.
---
>    * Load plugin manifest from the given URL.
193,194c305,307
<   async getPluginManifest(baseURL: string) {
<     const manifestURL = resolveURL(baseURL, PLUGIN_MANIFEST);
---
>   private async loadPluginManifest(manifestURL: string) {
>     const response = await this.options.fetchImpl(manifestURL, { cache: 'no-cache' });
>     const responseText = await response.text();
196c309,310
<     consoleLogger.info(`Loading plugin manifest from ${manifestURL}`);
---
>     return this.options.postProcessManifest(JSON.parse(responseText));
>   }
198,199c312,316
<     const response = await this.options.fetchImpl(manifestURL);
<     const responseText = await response.text();
---
>   /**
>    * Load plugin scripts from the given URL.
>    */
>   private async loadPluginScripts(baseURL: string, manifest: PluginManifest, data: PluginLoadData) {
>     const pluginName = manifest.name;
201,203c318,321
<     let manifest: PluginManifest = await pluginManifestSchema
<       .strict(true)
<       .validate(JSON.parse(responseText));
---
>     const [, rejectedReasons] = await settleAllPromises(
>       manifest.loadScripts.map((scriptName) => {
>         const scriptID = `${pluginName}/${scriptName}`;
>         const scriptElement = getScriptElement(scriptID);
205c323,325
<     manifest = await this.options.postProcessManifest(manifest);
---
>         if (scriptElement && !this.options.canReloadScript(manifest, scriptName)) {
>           return Promise.resolve();
>         }
207c327,382
<     return manifest;
---
>         if (scriptElement) {
>           scriptElement.remove();
>         }
> 
>         return injectScriptElement(
>           resolveURL(baseURL, scriptName, (url) => {
>             url.searchParams.set('cacheBuster', getRandomString());
>             return url;
>           }),
>           scriptID,
>         );
>       }),
>     );
> 
>     if (rejectedReasons.length > 0) {
>       throw new ErrorWithCause(
>         `Detected ${rejectedReasons.length} errors while loading plugin scripts`,
>         rejectedReasons,
>       );
>     }
> 
>     if (manifest.registrationMethod === 'callback' && !data.entryCallbackFired) {
>       data.status = 'failed';
>       this.invokeListeners({
>         success: false,
>         pluginName,
>         errorMessage: `Scripts of plugin ${pluginName} loaded without entry callback`,
>       });
>       return;
>     }
> 
>     if (manifest.registrationMethod === 'callback' && !data.entryCallbackModule) {
>       data.status = 'failed';
>       this.invokeListeners({
>         success: false,
>         pluginName,
>         errorMessage: `Entry callback for plugin ${pluginName} called without entry module`,
>       });
>       return;
>     }
> 
>     const entryModule =
>       manifest.registrationMethod === 'callback'
>         ? data.entryCallbackModule
>         : this.options.getPluginEntryModule(manifest);
> 
>     if (entryModule) {
>       await this.processPlugin(data, entryModule);
>     } else {
>       data.status = 'failed';
>       this.invokeListeners({
>         success: false,
>         pluginName,
>         errorMessage: `Failed to retrieve entry module of plugin ${pluginName}`,
>       });
>     }
210c385,421
<   private getDependencyResolutions() {
---
>   private processPlugin = async (data: PluginLoadData, entryModule: PluginEntryModule) => {
>     const { manifest } = data;
>     const pluginName = manifest.name;
> 
>     if (typeof entryModule.init !== 'function' || typeof entryModule.get !== 'function') {
>       data.status = 'failed';
>       this.invokeListeners({
>         success: false,
>         pluginName,
>         errorMessage: `Entry module of plugin ${pluginName} does not meet expected contract`,
>       });
>       return;
>     }
> 
>     try {
>       await Promise.resolve(entryModule.init(this.options.sharedScope));
>     } catch (e) {
>       data.status = 'failed';
>       this.invokeListeners({
>         success: false,
>         pluginName,
>         errorMessage: `Failed to initialize shared modules of plugin ${pluginName}`,
>         errorCause: e,
>       });
>       return;
>     }
> 
>     data.status = 'loaded';
>     this.invokeListeners({
>       success: true,
>       pluginName,
>       manifest,
>       entryModule,
>     });
>   };
> 
>   private getCurrentDependencyResolutions() {
235c446
<   async resolvePluginDependencies(manifest: PluginManifest) {
---
>   private async resolvePluginDependencies(manifest: PluginManifest) {
257c468
<         const resolutions = this.getDependencyResolutions();
---
>         const resolutions = this.getCurrentDependencyResolutions();
286d496
< 
308,355d517
<   /**
<    * Start loading the plugin entry script from `baseURL`.
<    */
<   loadPluginEntryScript(
<     baseURL: string,
<     manifest: PluginManifest,
<     getDocument: () => typeof document = _.constant(document),
<   ) {
<     const pluginName = manifest.name;
<     const data = this.plugins.get(pluginName);
< 
<     if (data?.status !== 'pending') {
<       consoleLogger.warn(`Attempt to load entry script of non-pending plugin ${pluginName}`);
<       return;
<     }
< 
<     const script = getDocument().createElement('script');
<     const scriptURL = resolveURL(baseURL, REMOTE_ENTRY_SCRIPT);
< 
<     script.src = scriptURL;
<     script.async = true;
< 
<     script.onload = () => {
<       if (!data.entryCallbackFired) {
<         data.status = 'failed';
<         this.invokeListeners({
<           success: false,
<           pluginName,
<           errorMessage: `Entry script of plugin ${pluginName} loaded without callback`,
<         });
<       }
<     };
< 
<     script.onerror = (event) => {
<       data.status = 'failed';
<       this.invokeListeners({
<         success: false,
<         pluginName,
<         errorMessage: `Failed to load entry script of plugin ${pluginName}`,
<         errorCause: event,
<       });
<     };
< 
<     consoleLogger.info(`Loading entry script of plugin ${pluginName} from ${scriptURL}`);
< 
<     getDocument().head.appendChild(script);
<   }
< 
359c521
<         consoleLogger.warn(`Received entry callback for unknown plugin ${pluginName}`);
---
>         consoleLogger.warn('Received entry callback for unknown plugin');
372,392c534
< 
<       Promise.resolve(entryModule.init(this.options.sharedScope))
<         // eslint-disable-next-line promise/always-return -- entryModule.init() returns Promise<void>
<         .then(() => {
<           data.status = 'loaded';
<           this.invokeListeners({
<             success: true,
<             pluginName,
<             manifest: data.manifest,
<             entryModule,
<           });
<         })
<         .catch((e) => {
<           data.status = 'failed';
<           this.invokeListeners({
<             success: false,
<             pluginName,
<             errorMessage: `Failed to initialize shared modules of plugin ${pluginName}`,
<             errorCause: e,
<           });
<         });
---
>       data.entryCallbackModule = entryModule;
397a540,541
>    *
>    * This must be called in order to load plugins using the `callback` registration method.
399,400c543
<   registerPluginEntryCallback(getWindow: () => typeof window = _.constant(window)) {
<     const windowGlobal = getWindow() as unknown as AnyObject;
---
>   registerPluginEntryCallback() {
403,404c546,547
<     if (typeof windowGlobal[callbackName] === 'function') {
<       consoleLogger.warn(`Plugin callback ${callbackName} is already registered`);
---
>     if (typeof window[callbackName] === 'function') {
>       consoleLogger.warn(`Plugin entry callback ${callbackName} is already registered`);
408c551
<     windowGlobal[callbackName] = this.createPluginEntryCallback();
---
>     window[callbackName] = this.createPluginEntryCallback();
```

Reply:
Sorry, this was my fault, I didn't do `git mv` on this file, so it shows up as removed and added like it was new.